File: Bundled_pipelines/mc_cdr3_mutations/mutation_frequency.r
#Using shazam to generate mutation frequency
#slancast@scg4.stanford.edu:/srv/gsfs0/projects/snyder/slancast/IMGT_analyses/IMGTcombined/analyses/mutation_freq
#

library(alakazam)
library(shazam)
library(ggplot2)

Subject <- Sys.getenv(c("var"))

print("T1")
T1 <- readChangeoDb(paste("../../combined_P",Subject,"_T1_db-pass_clone-pass_germ-pass.tab",sep=""))
ID <- c()
Isotype <- c()
for (i in 1:nrow(T1))  { ID <- append(ID,paste(Subject,"T1",sep="")) }
T1 <- cbind(T1, ID)

print("T2")
T2 <- readChangeoDb(paste("../../combined_P",Subject,"_T2_db-pass_clone-pass_germ-pass.tab",sep=""))
ID <- c()
Isotype <- c()
for (i in 1:nrow(T2))  { ID <- append(ID,paste(Subject,"T2",sep="")) }
T2 <- cbind(T2, ID)

print("T3")
T3 <- readChangeoDb(paste("../../combined_P",Subject,"_T3_db-pass_clone-pass_germ-pass.tab",sep=""))
ID <- c()
Isotype <- c()
for (i in 1:nrow(T3))  { ID <- append(ID,paste(Subject,"T3",sep="")) }
T3 <- cbind(T3, ID)

db <- rbind(T1, T2, T3)

db_new <- calcDBObservedMutations(db, sequenceColumn="SEQUENCE_IMGT", germlineColumn="GERMLINE_IMGT_D_MASK", frequency=FALSE,regionDefinition=IMGT_V_BY_REGIONS)
print("mutations_calculated")

print("db type")
print(typeof(db_new))

print(ncol(db_new))

write.table(db_new, file=paste(Subject,"db_new.csv", sep=""), sep=",")

pdf(paste("Mutation_frequency_",Subject,"_0FWR1.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_FWR1_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR1_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_0CDR1.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_CDR1_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR1_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_0FWR2.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_FWR2_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR2_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_0CDR2.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_CDR2_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR2_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()

pdf(paste("Mutation_frequency_",Subject,"_0FWR3.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_FWR3_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR3_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_0CDR3.pdf",sep=""))
ggplot(db_new, aes(x=ID, y=OBSERVED_CDR3_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR3_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


print("db type")
print(typeof(db_new))
CDR3_Mut <- matrix(nrow = 0,ncol = ncol(db_new))
colnames(CDR3_Mut) <- colnames(db_new)

print(CDR3_Mut)
print(ncol(db_new))
print("beginning for loop")


for (i in 1:nrow(db_new)){
if (as.integer(db_new[i,]$OBSERVED_CDR3_R) >= 1){
print(i)
CDR3_Mut <- rbind(CDR3_Mut, db_new[i,])  }}
write.table(CDR3_Mut, "CDR3_Mut.tab", sep="\t")
CDR3_Mut <- data.frame(CDR3_Mut)

print("making plots without 0 mutation")

pdf(paste("Mutation_frequency_",Subject,"_FWR1.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_FWR1_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR1_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_CDR1.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_CDR1_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR1_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_FWR2.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_FWR2_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR2_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_CDR2.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_CDR2_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR2_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()
    
pdf(paste("Mutation_frequency_",Subject,"_FWR3.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_FWR3_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_FWR3_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()


pdf(paste("Mutation_frequency_",Subject,"_CDR3.pdf",sep=""))
ggplot(CDR3_Mut, aes(x=ID, y=OBSERVED_CDR3_R)) +
	geom_violin(aes(fill=ID)) +
    ggtitle("OBSERVED_CDR3_R") + 
    xlab("ID") + 
    ylab("Mutation_Frequency") +
    theme(panel.background = element_rect(fill = "white"))
dev.off()

File: Bundled_pipelines/mc_cdr3_mutations/mutation_frequency.py
#slancast@scg4.stanford.edu:/srv/gsfs0/projects/snyder/slancast/IMGT_analyses/IMGTcombined/analyses/mutation_freq

import pandas as pd
import os
import numpy as np

print(str(os.environ['var']))

df = pd.read_csv(str(os.environ['var'])+"db_new.csv")

df = pd.DataFrame(df)
 
df["Total_Mutations"] = df["OBSERVED_FWR1_R"] + df["OBSERVED_CDR1_R"] + df["OBSERVED_FWR2_R"] + df["OBSERVED_CDR2_R"] + df["OBSERVED_FWR3_R"] + df["OBSERVED_CDR3_R"]

df.to_csv(str(os.environ['var'])+"CDR3_Mut.csv")

nucs = open("nucs_in_most_common_CDR3_"+str(os.environ['var'])+"t1.csv","r").readlines()

nucs[0] = nucs[0].split(",")
nucs[1] = nucs[1].split(",")
nucs[2] = nucs[2].split(",")
nucs[3] = nucs[3].split(",")

#print(nucs[0])

df = df.values.tolist()

mutations_CDR3_1 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_2 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_3 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_4 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]

total_mutations_CDR3_1 = []
total_mutations_CDR3_2 = []
total_mutations_CDR3_3 = []
total_mutations_CDR3_4 = []
 
print("adding mutations")
for entry in df:
	if len(entry) > 60:
		if entry[1] in nucs[0]:
			total_mutations_CDR3_1.append(entry[60])
			mutations_CDR3_1.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[1]:
			total_mutations_CDR3_2.append(entry[60])
			mutations_CDR3_2.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[2]:
			total_mutations_CDR3_3.append(entry[60])
			mutations_CDR3_3.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[3]:
			total_mutations_CDR3_4.append(entry[60])	
			mutations_CDR3_4.append(entry[48:59])

CDR3_1 = np.array(mutations_CDR3_1)
CDR3_2 = np.array(mutations_CDR3_2)
CDR3_3 = np.array(mutations_CDR3_3)
CDR3_4 = np.array(mutations_CDR3_4)

np.savetxt("mutations_in_most_common_CDR3_1_p"+str(os.environ['var'])+"t1.csv", CDR3_1, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_2_p"+str(os.environ['var'])+"t1.csv", CDR3_2, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_3_p"+str(os.environ['var'])+"t1.csv", CDR3_3, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_4_p"+str(os.environ['var'])+"t1.csv", CDR3_4, delimiter=",", fmt="%s")

print("Mean mutations in t1 CDR3 1:")
print(np.mean(total_mutations_CDR3_1))
print("Mean mutations in t1 CDR3 2:")
print(np.mean(total_mutations_CDR3_2))
print("Mean mutations in t1 CDR3 3:")
print(np.mean(total_mutations_CDR3_3))
print("Mean mutations in t1 CDR3 4:")
print(np.mean(total_mutations_CDR3_4))

##################################################################################
#																				 #
#																				 #
#										T2										 #
#																				 #
#																				 #
##################################################################################	

nucs = open("nucs_in_most_common_CDR3_"+str(os.environ['var'])+"t2.csv","r").readlines()

nucs[0] = nucs[0].split(",")
nucs[1] = nucs[1].split(",")
nucs[2] = nucs[2].split(",")
nucs[3] = nucs[3].split(",")

mutations_CDR3_1 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_2 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_3 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_4 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]

total_mutations_CDR3_1 = []
total_mutations_CDR3_2 = []
total_mutations_CDR3_3 = []
total_mutations_CDR3_4 = []
 
print("adding mutations")
for entry in df:
	if len(entry) > 60:
		if entry[1] in nucs[0]:
			total_mutations_CDR3_1.append(entry[60])
			mutations_CDR3_1.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[1]:
			total_mutations_CDR3_2.append(entry[60])
			mutations_CDR3_2.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[2]:
			total_mutations_CDR3_3.append(entry[60])
			mutations_CDR3_3.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[3]:
			total_mutations_CDR3_4.append(entry[60])	
			mutations_CDR3_4.append(entry[48:59])

CDR3_1 = np.array(mutations_CDR3_1)
CDR3_2 = np.array(mutations_CDR3_2)
CDR3_3 = np.array(mutations_CDR3_3)
CDR3_4 = np.array(mutations_CDR3_4)

np.savetxt("mutations_in_most_common_CDR3_1_p"+str(os.environ['var'])+"t2.csv", CDR3_1, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_2_p"+str(os.environ['var'])+"t2.csv", CDR3_2, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_3_p"+str(os.environ['var'])+"t2.csv", CDR3_3, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_4_p"+str(os.environ['var'])+"t2.csv", CDR3_4, delimiter=",", fmt="%s")

print("Mean mutations in t2 CDR3 1:")
print(np.mean(total_mutations_CDR3_1))
print("Mean mutations in t2 CDR3 2:")
print(np.mean(total_mutations_CDR3_2))
print("Mean mutations in t2 CDR3 3:")
print(np.mean(total_mutations_CDR3_3))
print("Mean mutations in t2 CDR3 4:")
print(np.mean(total_mutations_CDR3_4))

##################################################################################
#																				 #
#																				 #
#										T3										 #
#																				 #
#																				 #
##################################################################################	

nucs = open("nucs_in_most_common_CDR3_"+str(os.environ['var'])+"t3.csv","r").readlines()

nucs[0] = nucs[0].split(",")
nucs[1] = nucs[1].split(",")
nucs[2] = nucs[2].split(",")
nucs[3] = nucs[3].split(",")

mutations_CDR3_1 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_2 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_3 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]
mutations_CDR3_4 = [["OBSERVED_CDR1_R","OBSERVED_CDR1_S","OBSERVED_CDR2_R","OBSERVED_CDR2_S","OBSERVED_CDR3_R","OBSERVED_CDR3_S","OBSERVED_FWR1_R","OBSERVED_FWR1_S","OBSERVED_FWR2_R","OBSERVED_FWR2_S","OBSERVED_FWR3_R","OBSERVED_FWR3_S"]]

total_mutations_CDR3_1 = []
total_mutations_CDR3_2 = []
total_mutations_CDR3_3 = []
total_mutations_CDR3_4 = []
 
print("adding mutations")
for entry in df:
	if len(entry) > 60:
		if entry[1] in nucs[0]:
			total_mutations_CDR3_1.append(entry[60])
			mutations_CDR3_1.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[1]:
			total_mutations_CDR3_2.append(entry[60])
			mutations_CDR3_2.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[2]:
			total_mutations_CDR3_3.append(entry[60])
			mutations_CDR3_3.append(entry[48:59])
	if len(entry) > 60:
		if entry[1] in nucs[3]:
			total_mutations_CDR3_4.append(entry[60])	
			mutations_CDR3_4.append(entry[48:59])

CDR3_1 = np.array(mutations_CDR3_1)
CDR3_2 = np.array(mutations_CDR3_2)
CDR3_3 = np.array(mutations_CDR3_3)
CDR3_4 = np.array(mutations_CDR3_4)

np.savetxt("mutations_in_most_common_CDR3_1_p"+str(os.environ['var'])+"t3.csv", CDR3_1, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_2_p"+str(os.environ['var'])+"t3.csv", CDR3_2, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_3_p"+str(os.environ['var'])+"t3.csv", CDR3_3, delimiter=",", fmt="%s")
np.savetxt("mutations_in_most_common_CDR3_4_p"+str(os.environ['var'])+"t3.csv", CDR3_4, delimiter=",", fmt="%s")

print("Mean mutations in t3 CDR3 1:")
print(np.mean(total_mutations_CDR3_1))
print("Mean mutations in t3 CDR3 2:")
print(np.mean(total_mutations_CDR3_2))
print("Mean mutations in t3 CDR3 3:")
print(np.mean(total_mutations_CDR3_3))
print("Mean mutations in t3 CDR3 4:")
print(np.mean(total_mutations_CDR3_4))